### DISCUSSION TITLE
mqtt verliert verbindung

### DISCUSSION THREAD

**Discussion by Ivangrosnij** [tbnobody/OpenDTU#533] [General]
Hallo,

ich betreibe HM-1500 mit OpenDTU auf ESP32 mit ioBroker

Letzte zeit verliert OpenDTU mqtt verbindung zum ioBroker. 

Nach eine neustart von OpenDTU verbindet er sich wieder.

Interesant ist auch das über Weboberläche Info/MQTT/Verbindungsstatus auf getrennt ist.
und in ioBroker /DTU/Status/ online zeigt.

**Comment by tbnobody** [comment]
Hallo,

das ist sehr komisch. Das "offline" auf dem Statustopic ist eigentlich eine LWT Message. Das bedeutet, dass diese bei einem Disconnect des Clients (auch wenn dieser hart ausgeschaltet wird) vom Broker (Server) gesetzt wird. Der Client schickt nie ein "offline". 

Was passiert wenn du im Fehlerfall (also wenn OpenDTU die Verbindung verliert) du den ESP aussteckst? Setzt wir dann das Topic auf "offline" gesetzt? Falls nein liegt der Fehler am Broker (Server).

**Reply by Ivangrosnij** [reply]
Das probiere ich beim nächsten Fehlerall.

**Comment by cbeham1** [comment]
Guten Tag,

was war die Lösung. Habe das auch gerade.

Danke

**Comment by Ivangrosnij** [comment]
Hi, ich habe das leider noch nicht lösen können das passiert nur nach dem man iobroker neugestartet hat. Danach muss man OpenDTU neustarten danach steht die verbiendung wieder.

**Comment by tbnobody** [comment]
Habe es gerade versucht nochmal nachzustellen.
* Mosquitto beendet, 5min gewartet, wieder gestartet --> Sauberer Reconnect von OpenDTU
* Mosquitto mit sigkill beendet, 5min gewartet, wieder gestartet --> Sauberer Reconnect von OpenDTU
* Netzwerkkabel zum Broker gezogen, nach ca. 30sek sieht man in der Console (man muss wirklich die Serielle verwenden) von OpenDTU 
eine Disconnect meldung und es wird zyklisch ein Reconnect versucht. Wenn man das Kabel wieder einsteckt --> Sauberer Reconnect von OpenDTU

Ich weiß nicht was ich hier noch testen soll. Könnte mir auch vorstellen das der in ioBroker integrierte MQTT Broker seltsame dinge tut.

@Ivangrosnij würdest du das ggf. mal auf den Output der Seriellen Konsole achten während du den ioBroker neustartest?

**Comment by Ivangrosnij** [comment]
ich habe momentan OpenDTU-OnBattery drauf laufen da ist was MQTT angeht gleich wenn man ioBroker neustartet geht die Verbiendung verloren.
hier ist der log aus der Konsole
`22:45:44.288 > RX Period End
22:45:44.289 > All missing
22:45:44.289 > Nothing received, resend whole request
22:45:44.289 > TX SystemConfigPara Channel: 40 --> 15 84 00 54 42 80 15 51 16 80 05 00 64 1C C8 7E 00 00 00 00 00 00 00 00 AF 19 A8 
22:45:44.591 > RX Period End
22:45:44.591 > All missing
22:45:44.591 > Nothing received, resend whole request
22:45:44.591 > TX SystemConfigPara Channel: 61 --> 15 84 00 54 42 80 15 51 16 80 05 00 64 1C C8 7E 00 00 00 00 00 00 00 00 AF 19 A8 
22:45:45.124 > RX Period End
22:45:45.124 > All missing
22:45:45.124 > Nothing received, resend whole request
22:45:45.124 > TX SystemConfigPara Channel: 75 --> 15 84 00 54 42 80 15 51 16 80 05 00 64 1C C8 7E 00 00 00 00 00 00 00 00 AF 19 A8 
22:45:45.411 > RX Period End
22:45:45.411 > All missing
22:45:45.411 > Nothing received, resend count exeeded
22:45:45.718 > Received MQTT message on topic: shellies/shellyem3-xxxxxxxxxxxxx/emeter/0/power
22:45:45.718 > PowerLimiterClass: Received MQTT message on topic: shellies/shellyem3-xxxxxxxxxxxxx/emeter/0/power
22:45:46.028 > Received MQTT message on topic: shellies/shellyem3-xxxxxxxxxxxxx/emeter/1/power
22:45:46.028 > PowerLimiterClass: 22:45:46.330 > Received MQTT message on topic: shellies/shellyem3-xxxxxxxxxxxxx/emeter/1/power
22:45:46.330 > Received MQTT message on topic: shellies/shellyem3-xxxxxxxxxxxxx/emeter/2/power
22:45:46.330 > PowerLimiterClass: Recei22:45:46.745 > ved MQTT message on topic: shellies/shellyem3-xxxxxxxxxxxxx/emeter/2/power
22:45:47.050 > Received MQTT message on topic: shellies/shellyem3-xxxxxxxxxxxxx/emeter/0/power
22:45:47.050 > PowerLim22:45:47.356 > iterClass: Received MQTT message on topic: shellies/shellyem3-xxxxxxxxxxxxx/emeter/0/power
22:45:47.576 > Received MQTT message on topic: shellies/shellyem3-xxxxxxxxxxxxx/emeter/1/power
22:45:47.577 > PowerLimiterClass: Received MQTT message on topic: shellies/shellyem3-xxxxxxxxxxxxx/emeter/1/power
22:45:47.577 > Recei22:45:47.868 > ved MQTT message on topic: shellies/shellyem3-xxxxxxxxxxxxx/emeter/2/power
22:45:47.868 > PowerLimiterClass: Received MQTT message on topic: shellies/shellyem3-xxxxxxxxxxxxx/emeter22:45:48.180 > /2/power
22:45:48.385 > Websocket: [/livedata][5] disconnect
22:45:48.596 > Disconnected from MQTT.
22:45:48.596 > Disconnect reason:TCP_DISCONNECTED
22:45:48.900 > Fetch inverter: "ID Inverter"
22:45:48.900 > Request SystemConfigPara
22:45:50.597 > C22:45:50.839 > onnecting to MQTT..22:45:51.145 > .
22:45:53.905 > Fetch inverter: "ID Inverter"
22:45:53.905 > Request SystemConfigPara
22:45:54.222 > TX RealTimeRunData Channel: 3 --> 15 84 00 54 42 80 15 51 16 80 0B 00 64 1C C8 8D 00 00 00 00 00 00 00 00 95 06 70 
22:45:54.526 > RX Period End
22:45:54.526 > All missing
22:45:54.526 > Nothing received, resend whole request
22:45:54.526 > TX RealTimeRunData Channel: 23 --> 15 84 00 54 42 80 15 51 16 80 0B 00 64 1C C8 8D 00 00 00 00 00 00 00 00 95 06 70 
22:45:54.833 > RX Period End
22:45:54.833 > All missing
22:45:54.833 > Nothing received, resend whole request
22:45:54.833 > TX RealTimeRunData Channel: 40 --> 15 84 00 54 42 80 15 51 16 80 0B 00 64 1C C8 8D 00 00 00 00 00 00 00 00 95 06 70 
22:45:55.138 > RX Period End
22:45:55.138 > All missing
22:45:55.138 > Nothing received, resend whole request
22:45:55.138 > TX RealTimeRunData Channel: 61 --> 15 84 00 54 42 80 15 51 16 80 0B 00 64 1C C8 8D 00 00 00 00 00 00 00 00 95 06 70 
22:45:55.445 > RX Period End
22:45:55.445 > All missing
22:45:55.445 > Nothing received, resend whole request
22:45:55.445 > TX RealTimeRunData Channel: 75 --> 15 84 00 54 42 80 15 51 16 80 0B 00 64 1C C8 8D 00 00 00 00 00 00 00 00 95 06 70 
22:45:55.753 > RX Period End
22:45:55.753 > All missing
22:45:55.753 > Nothing received, resend count exeeded
22:45:56.060 > TX AlarmData Channel: 3 --> 15 84 00 54 42 80 15 51 16 80 11 00 64 1C C8 8D 00 00 00 00 00 00 00 00 4F 1D AB 
22:45:56.367 > RX Period End
22:45:56.367 > All missing
22:45:56.367 > Nothing received, resend whole request
22:45:56.367 > TX AlarmData Channel: 23 --> 15 84 00 54 42 80 15 51 16 80 11 00 64 1C C8 8D 00 00 00 00 00 00 00 00 4F 1D AB 
22:45:56.675 > RX Period End
22:45:56.675 > All missing
22:45:56.675 > Nothing received, resend whole request
22:45:56.675 > TX AlarmData Channel: 40 --> 15 84 00 54 42 80 15 51 16 80 11 00 64 1C C8 8D 00 00 00 00 00 00 00 00 4F 1D AB 
22:45:57.106 > RX Period End
22:45:57.106 > All missing
22:45:57.106 > Nothing received, resend whole request
22:45:57.106 > TX AlarmData Channel: 61 --> 15 84 00 54 42 80 15 51 16 80 11 00 64 1C C8 8D 00 00 00 00 00 00 00 00 4F 1D AB 
22:45:57.748 > RX Period End
22:45:57.748 > All missing
22:45:57.748 > Nothing received, resend whole request
22:45:57.748 > TX AlarmData Channel: 75 --> 15 84 00 54 42 80 15 51 16 80 11 00 64 1C C8 8D 00 00 00 00 00 00 00 00 4F 1D AB 
22:45:58.354 > RX Period End
22:45:58.354 > All missing
22:45:58.354 > Nothing received, resend count exeeded
22:45:58.723 > TX SystemConfigPara Channel: 3 --> 15 84 00 54 42 80 15 51 16 80 05 00 64 1C C8 8D 00 00 00 00 00 00 00 00 5B 09 BF 
22:45:59.030 > RX Period End
22:45:59.031 > All missing
22:45:59.031 > Nothing received, resend whole request
22:45:59.031 > TX SystemConfigPara Channel: 23 --> 15 84 00 54 42 80 15 51 16 80 05 00 64 1C C8 8D 00 00 00 00 00 00 00 00 5B 09 BF 
22:45:59.336 > RX Period End
22:45:59.336 > All missing
22:45:59.336 > Nothing received, resend whole request
22:45:59.336 > TX SystemConfigPara Channel: 40 --> 15 84 00 54 42 80 15 51 16 80 05 00 64 1C C8 8D 00 00 00 00 00 00 00 00 5B 09 BF 
22:45:59.645 > RX Period End
22:45:59.645 > All missing
22:45:59.645 > Nothing received, resend whole request
22:45:59.645 > TX SystemConfigPara Channel: 61 --> 15 84 00 54 42 80 15 51 16 80 05 00 64 1C C8 8D 00 00 00 00 00 00 00 00 5B 09 BF 
22:45:59.954 > RX Period End
22:45:59.954 > All missing
22:45:59.954 > Nothing received, resend whole request
22:45:59.954 > TX SystemConfigPara Channel: 75 --> 15 84 00 54 42 80 15 51 16 80 05 00 64 1C C8 8D 00 00 00 00 00 00 00 00 5B 09 BF 
22:46:00.262 > RX Period End
22:46:00.262 > All missing
22:46:00.262 > Nothing received, resend count exeeded
22:46:03.918 > Fetch inverter: "ID Inverter"
22:46:03.918 > Request SystemConfigPara
22:46:08.929 > Fetch inverter: "ID Inverter"
22:46:08.929 > Request SystemConfigPara
22:46:09.270 > TX RealTimeRunData Channel: 3 --> 15 84 00 54 42 80 15 51 16 80 0B 00 64 1C C8 9C 00 00 00 00 00 00 00 00 C5 C6 F1 
22:46:09.583 > RX Period End
22:46:09.583 > All missing
22:46:09.583 > Nothing received, resend whole request
22:46:09.583 > TX RealTimeRunData Channel: 23 --> 15 84 00 54 42 80 15 51 16 80 0B 00 64 1C C8 9C 00 00 00 00 00 00 00 00 C5 C6 F1 
22:46:09.884 > RX Period End
22:46:09.884 > All missing
22:46:09.884 > Nothing received, resend whole request
22:46:09.884 > TX RealTimeRunData Channel: 40 --> 15 84 00 54 42 80 15 51 16 80 0B 00 64 1C C8 9C 00 00 00 00 00 00 00 00 C5 C6 F1 
22:46:10.194 > RX Period End
22:46:10.194 > All missing
22:46:10.194 > Nothing received, resend whole request
22:46:10.194 > TX RealTimeRunData Channel: 61 --> 15 84 00 54 42 80 15 51 16 80 0B 00 64 1C C8 9C 00 00 00 00 00 00 00 00 C5 C6 F1 
22:46:10.502 > RX Period End
22:46:10.502 > All missing
22:46:10.502 > Nothing received, resend whole request
22:46:10.502 > TX RealTimeRunData Channel: 75 --> 15 84 00 54 42 80 15 51 16 80 0B 00 64 1C C8 9C 00 00 00 00 00 00 00 00 C5 C6 F1 
22:46:10.804 > RX Period End
22:46:10.804 > All missing
22:46:10.804 > Nothing received, resend count exeeded
22:46:11.113 > TX AlarmData Channel: 3 --> 15 84 00 54 42 80 15 51 16 80 11 00 64 1C C8 9C 00 00 00 00 00 00 00 00 1F DD 2A 
22:46:11.421 > RX Period End
22:46:11.421 > All missing
22:46:11.421 > Nothing received, resend whole request
22:46:11.421 > TX AlarmData Channel: 23 --> 15 84 00 54 42 80 15 51 16 80 11 00 64 1C C8 9C 00 00 00 00 00 00 00 00 1F DD 2A 
22:46:11.727 > RX Period End
22:46:11.727 > All missing
22:46:11.727 > Nothing received, resend whole request
22:46:11.727 > TX AlarmData Channel: 40 --> 15 84 00 54 42 80 15 51 16 80 11 00 64 1C C8 9C 00 00 00 00 00 00 00 00 1F DD 2A 
22:46:12.113 > RX Period End
22:46:12.113 > All missing
22:46:12.113 > Nothing received, resend whole request
22:46:12.113 > TX AlarmData Channel: 61 --> 15 84 00 54 42 80 15 51 16 80 11 00 64 1C C8 9C 00 00 00 00 00 00 00 00 1F DD 2A 
22:46:12.753 > RX Period End
22:46:12.753 > All missing
22:46:12.753 > Nothing received, resend whole request
22:46:12.753 > TX AlarmData Channel: 75 --> 15 84 00 54 42 80 15 51 16 80 11 00 64 1C C8 9C 00 00 00 00 00 00 00 00 1F DD 2A 
22:46:13.359 > RX Period End
22:46:13.359 > All missing
22:46:13.359 > Nothing received, resend count exeeded
22:46:13.674 > TX SystemConfigPara Channel: 3 --> 15 84 00 54 42 80 15 51 16 80 05 00 64 1C C8 9C 00 00 00 00 00 00 00 00 0B C9 3E 
22:46:13.981 > RX Period End
22:46:13.981 > All missing
22:46:13.981 > Nothing received, resend whole request
22:46:13.981 > TX SystemConfigPara Channel: 23 --> 15 84 00 54 42 80 15 51 16 80 05 00 64 1C C8 9C 00 00 00 00 00 00 00 00 0B C9 3E 
22:46:14.291 > RX Period End
22:46:14.291 > All missing
22:46:14.291 > Nothing received, resend whole request
22:46:14.291 > TX SystemConfigPara Channel: 40 --> 15 84 00 54 42 80 15 51 16 80 05 00 64 1C C8 9C 00 00 00 00 00 00 00 00 0B C9 3E 
22:46:14.598 > RX Period End
22:46:14.598 > All missing
22:46:14.598 > Nothing received, resend whole request
22:46:14.598 > TX SystemConfigPara Channel: 61 --> 15 84 00 54 42 80 15 51 16 80 05 00 64 1C C8 9C 00 00 00 00 00 00 00 00 0B C9 3E 
22:46:14.904 > RX Period End
22:46:14.904 > All missing
22:46:14.904 > Nothing received, resend whole request
22:46:14.904 > TX SystemConfigPara Channel: 75 --> 15 84 00 54 42 80 15 51 16 80 05 00 64 1C C8 9C 00 00 00 00 00 00 00 00 0B C9 3E 
22:46:15.209 > RX Period End
22:46:15.209 > All missing
22:46:15.209 > Nothing received, resend count exeeded
22:46:18.913 > Fetch inverter: "ID Inverter"
22:46:18.913 > Request SystemConfigPara
22:46:23.921 > Fetch inverter: "ID Inverter"
22:46:23.921 > Request SystemConfigPara
22:46:24.126 > TX RealTimeRunData Channel: 3 --> 15 84 00 54 42 80 15 51 16 80 0B 00 64 1C C8 AB 00 00 00 00 00 00 00 00 F4 B4 85 
22:46:24.425 > RX Period End
22:46:24.425 > All missing
22:46:24.425 > Nothing received, resend whole request
22:46:24.425 > TX RealTimeRunData Channel: 23 --> 15 84 00 54 42 80 15 51 16 80 0B 00 64 1C C8 AB 00 00 00 00 00 00 00 00 F4 B4 85 
22:46:24.738 > RX Period End
22:46:24.738 > All missing
22:46:24.738 > Nothing received, resend whole request
22:46:24.738 > TX RealTimeRunData Channel: 40 --> 15 84 00 54 42 80 15 51 16 80 0B 00 64 1C C8 AB 00 00 00 00 00 00 00 00 F4 B4 85 
22:46:25.039 > RX Period End
22:46:25.039 > All missing
22:46:25.039 > Nothing received, resend whole request
22:46:25.039 > TX RealTimeRunData Channel: 61 --> 15 84 00 54 42 80 15 51 16 80 0B 00 64 1C C8 AB 00 00 00 00 00 00 00 00 F4 B4 85 
22:46:25.351 > RX Period End
22:46:25.351 > All missing
22:46:25.351 > Nothing received, resend whole request
22:46:25.351 > TX RealTimeRunData Channel: 75 --> 15 84 00 54 42 80 15 51 16 80 0B 00 64 1C C8 AB 00 00 00 00 00 00 00 00 F4 B4 85 
22:46:25.654 > RX Period End
22:46:25.654 > All missing
22:46:25.654 > Nothing received, resend count exeeded
22:46:25.977 > TX AlarmData Channel: 3 --> 15 84 00 54 42 80 15 51 16 80 11 00 64 1C C8 AB 00 00 00 00 00 00 00 00 2E AF 5E 
22:46:26.270 > RX Period End
22:46:26.270 > All missing
22:46:26.270 > Nothing received, resend whole request
22:46:26.270 > TX AlarmData Channel: 23 --> 15 84 00 54 42 80 15 51 16 80 11 00 64 1C C8 AB 00 00 00 00 00 00 00 00 2E AF 5E 
22:46:26.577 > RX Period End
22:46:26.577 > All missing
22:46:26.577 > Nothing received, resend whole request
22:46:26.577 > TX AlarmData Channel: 40 --> 15 84 00 54 42 80 15 51 16 80 11 00 64 1C C8 AB 00 00 00 00 00 00 00 00 2E AF 5E 
22:46:27.118 > RX Period End
22:46:27.118 > All missing
22:46:27.118 > Nothing received, resend whole request
22:46:27.118 > TX AlarmData Channel: 61 --> 15 84 00 54 42 80 15 51 16 80 11 00 64 1C C8 AB 00 00 00 00 00 00 00 00 2E AF 5E 
22:46:27.758 > RX Period End
22:46:27.758 > All missing
22:46:27.758 > Nothing received, resend whole request
22:46:27.758 > TX AlarmData Channel: 75 --> 15 84 00 54 42 80 15 51 16 80 11 00 64 1C C8 AB 00 00 00 00 00 00 00 00 2E AF 5E 
22:46:28.360 > RX Period End
22:46:28.360 > All missing
22:46:28.360 > Nothing received, resend count exeeded
22:46:28.631 > TX SystemConfigPara Channel: 3 --> 15 84 00 54 42 80 15 51 16 80 05 00 64 1C C8 AB 00 00 00 00 00 00 00 00 3A BB 4A 
22:46:28.930 > RX Period End
22:46:28.930 > All missing
22:46:28.930 > Nothing received, resend whole request
22:46:28.930 > TX SystemConfigPara Channel: 23 --> 15 84 00 54 42 80 15 51 16 80 05 00 64 1C C8 AB 00 00 00 00 00 00 00 00 3A BB 4A 
22:46:29.239 > RX Period End
22:46:29.239 > All missing
22:46:29.239 > Nothing received, resend whole request
22:46:29.239 > TX SystemConfigPara Channel: 40 --> 15 84 00 54 42 80 15 51 16 80 05 00 64 1C C8 AB 00 00 00 00 00 00 00 00 3A BB 4A 
22:46:29.549 > RX Period End
22:46:29.549 > All missing
22:46:29.549 > Nothing received, resend whole request
22:46:29.549 > TX SystemConfigPara Channel: 61 --> 15 84 00 54 42 80 15 51 16 80 05 00 64 1C C8 AB 00 00 00 00 00 00 00 00 3A BB 4A 
22:46:29.851 > RX Period End
22:46:29.851 > All missing
22:46:29.851 > Nothing received, resend whole request
22:46:29.851 > TX SystemConfigPara Channel: 75 --> 15 84 00 54 42 80 15 51 16 80 05 00 64 1C C8 AB 00 00 00 00 00 00 00 00 3A BB 4A 
22:46:30.158 > RX Period End
22:46:30.158 > All missing
22:46:30.158 > Nothing received, resend count exeeded
22:46:33.921 > Fetch inverter: "ID Inverter"
22:46:33.921 > Request SystemConfigPara
22:46:38.923 > Fetch inverter: "ID Inverter"
22:46:38.924 > Request SystemConfigPara
22:46:39.170 > TX RealTimeRunData Channel: 3 --> 15 84 00 54 42 80 15 51 16 80 0B 00 64 1C C8 BA 00 00 00 00 00 00 00 00 A4 74 04 
22:46:39.478 > RX Period End
22:46:39.478 > All missing
22:46:39.478 > Nothing received, resend whole request
22:46:39.478 > TX RealTimeRunData Channel: 23 --> 15 84 00 54 42 80 15 51 16 80 0B 00 64 1C C8 BA 00 00 00 00 00 00 00 00 A4 74 04 
22:46:39.783 > RX Period End
22:46:39.783 > All missing
22:46:39.783 > Nothing received, resend whole request
22:46:39.783 > TX RealTimeRunData Channel: 40 --> 15 84 00 54 42 80 15 51 16 80 0B 00 64 1C C8 BA 00 00 00 00 00 00 00 00 A4 74 04 
22:46:40.093 > RX Period End
22:46:40.093 > All missing
22:46:40.093 > Nothing received, resend whole request
22:46:40.093 > TX RealTimeRunData Channel: 61 --> 15 84 00 54 42 80 15 51 16 80 0B 00 64 1C C8 BA 00 00 00 00 00 00 00 00 A4 74 04 
22:46:40.399 > RX Period End
22:46:40.400 > All missing
22:46:40.400 > Nothing received, resend whole request
22:46:40.400 > TX RealTimeRunData Channel: 75 --> 15 84 00 54 42 80 15 51 16 80 0B 00 64 1C C8 BA 00 00 00 00 00 00 00 00 A4 74 04 
22:46:40.705 > RX Period End
22:46:40.705 > All missing
22:46:40.705 > Nothing received, resend count exeeded
22:46:41.013 > TX AlarmData Channel: 3 --> 15 84 00 54 42 80 15 51 16 80 11 00 64 1C C8 BA 00 00 00 00 00 00 00 00 7E 6F DF 
22:46:41.422 > RX Period End
22:46:41.422 > All missing
22:46:41.422 > Nothing received, resend whole request
22:46:41.422 > TX AlarmData Channel: 23 --> 15 84 00 54 42 80 15 51 16 80 11 00 64 1C C8 BA 00 00 00 00 00 00 00 00 7E 6F DF 
22:46:41.731 > RX Period End
22:46:41.732 > All missing
22:46:41.732 > Nothing received, resend whole request
22:46:41.732 > TX AlarmData Channel: 40 --> 15 84 00 54 42 80 15 51 16 80 11 00 64 1C C8 BA 00 00 00 00 00 00 00 00 7E 6F DF 
22:46:42.120 > RX Period End
22:46:42.120 > All missing
22:46:42.120 > Nothing received, resend whole request
22:46:42.120 > TX AlarmData Channel: 61 --> 15 84 00 54 42 80 15 51 16 80 11 00 64 1C C8 BA 00 00 00 00 00 00 00 00 7E 6F DF 
22:46:42.765 > RX Period End
22:46:42.765 > All missing
22:46:42.765 > Nothing received, resend whole request
22:46:42.765 > TX AlarmData Channel: 75 --> 15 84 00 54 42 80 15 51 16 80 11 00 64 1C C8 BA 00 00 00 00 00 00 00 00 7E 6F DF 
22:46:43.366 > RX Period End
22:46:43.366 > All missing
22:46:43.366 > Nothing received, resend count exeeded
22:46:43.682 > TX SystemConfigPara Channel: 3 --> 15 84 00 54 42 80 15 51 16 80 05 00 64 1C C8 BA 00 00 00 00 00 00 00 00 6A 7B CB 
22:46:43.982 > RX Period End
22:46:43.982 > All missing
22:46:43.982 > Nothing received, resend whole request
22:46:43.982 > TX SystemConfigPara Channel: 23 --> 15 84 00 54 42 80 15 51 16 80 05 00 64 1C C8 BA 00 00 00 00 00 00 00 00 6A 7B CB 
22:46:44.290 > RX Period End
22:46:44.290 > All missing
22:46:44.290 > Nothing received, resend whole request
22:46:44.290 > TX SystemConfigPara Channel: 40 --> 15 84 00 54 42 80 15 51 16 80 05 00 64 1C C8 BA 00 00 00 00 00 00 00 00 6A 7B CB 
22:46:44.598 > RX Period End
22:46:44.598 > All missing
22:46:44.598 > Nothing received, resend whole request
22:46:44.598 > TX SystemConfigPara Channel: 61 --> 15 84 00 54 42 80 15 51 16 80 05 00 64 1C C8 BA 00 00 00 00 00 00 00 00 6A 7B CB 
22:46:44.906 > RX Period End
22:46:44.906 > All missing
22:46:44.906 > Nothing received, resend whole request
22:46:44.906 > TX SystemConfigPara Channel: 75 --> 15 84 00 54 42 80 15 51 16 80 05 00 64 1C C8 BA 00 00 00 00 00 00 00 00 6A 7B CB 
22:46:45.212 > RX Period End
22:46:45.212 > All missing
22:46:45.212 > Nothing received, resend count exeeded
`
zur Info
ich habe mit ioBroker mehrere Geräte per MQTT verbunden Shelly, mclighting da bleiben die verbindungen bestehen.

**Comment by tbnobody** [comment]
> ich habe mit ioBroker mehrere Geräte per MQTT verbunden Shelly, mclighting da bleiben die verbindungen bestehen.

wenn du den in ioBroker integrierten MQTT Broker verwendest bleiben die Verbindungen auch dort nicht bestehen. Sie werden nur wieder aufgebaut. 

Leider bringt ein Auszug aus der Console im Webinterface nichts da tiefgreifende Logs aus dem TCP Stack nur auf der richtigen **Seriellen** Konsole ausgegeben werden. Und genau das wäre wichtig. Weil wie man im oberen Log sieht, wird 1x versucht eine Verbindung aufzubauen: `Connecting to MQTT`. Es erscheint jedoch nie ein `Connected to MQTT.`. D.h. hier geht wohl irgendwas in der MQTT Library schief.

**Comment by yippieyaray** [comment]
Hier mein Fehlerbeschreibung und Korrekturmöglichkeit in englisch.

Ever since I started using OpenDTU, I have been troubled by this annoying problem, which I have already found described as a openDTU problem on several places.

I've analyzed and corrected it in temp branch of async-mqtt-client.

Regarding the cause:
OpenDTU initiates a connect request via an external MQTT client software (https://github.com/marvinroger/async-mqtt-client) to the MQTT server (e.g., ioBroker). It expects either a successful response via onMqttConnect-Callback or, through onMqttDisconnect-Callback, an indication that the connection was interrupted or the connection attempt failed. In the case of an onMqttDisconnect call, a new attempt to connect is initiated by OpenDTU. If the call to onMqttDisconnect is absent, no reconnect occurs.

The problem now, however, is that the MQTT client code has a bug in the connection setup. Under certain circumstances, the method MqttClient::loop() may stuck in the second phase of the connection setup if the MQTT server e.g. ioBroker allows a socket connect but unfortunately does not send back a connect-acknowledge packet in situation like server shutdown or restart.

The faulty code in MqttClient.cpp:

```cpp
void MqttClient::loop() { 
  switch (_state) {
    case State::connectingTcp1:
      ...
      ...
      if (_useIp ? _transport->connect(_ip, _port) : _transport->connect(_host, _port)) {
        _setState(State::connectingTcp2);   <---- connection successful; set _state to connectingTcp2
      } else {
        _setState(State::disconnectingTcp1);
        _disconnectReason = DisconnectReason::TCP_DISCONNECTED;
        break;
      }
      // Falling through to speed up connecting on blocking transport 'connect' implementations
      [[fallthrough]];
    case State::connectingTcp2:
      if (_transport->connected()) {    <---- connected is false until the ioBroker sends a connect-acknowledge-packet
        _parser.reset();
        _setState(State::connectingMqtt);  <---- ioBroker does not send an acknowledge, the _state will never set to next step connectingMqtt
      }
      break;


Here are my modifications:

void MqttClient::loop() {
  switch (_state) {
    case State::disconnected:
      #if defined(ARDUINO_ARCH_ESP32)
      if (_useInternalTask == espMqttClientTypes::UseInternalTask::YES) {
        vTaskSuspend(_taskHandle);
      }
      #endif
      break;
    case State::connectingTcp1:
      if (_useIp ? _transport->connect(_ip, _port) : _transport->connect(_host, _port)) {
        _setState(State::connectingTcp2);
        _lastClientActivity = millis();           // <---- set connect time
      } else {
        _setState(State::disconnectingTcp1);
        _disconnectReason = DisconnectReason::TCP_DISCONNECTED;
        break;
      }
      // Falling through to speed up connecting on blocking transport 'connect' implementations
      [[fallthrough]];
    case State::connectingTcp2:
      if (_transport->connected()) {
        _parser.reset();
        _lastClientActivity = _lastServerActivity = millis();
        _setState(State::connectingMqtt);
      }  else if (millis() - _lastClientActivity > _keepAlive) {  //  <---- new else block; use _keepAlive 15 seconds as timeout here 
        _setState(State::disconnectingTcp1);                         // <---- force an disconnect sequence that calls onMqttDisconnect which starts a new connect request
        _disconnectReason = DisconnectReason::TCP_DISCONNECTED;
      }
      break;
```

https://github.com/marvinroger/async-mqtt-client seems to be dead so I do not expect any fixes from this site.

Regards!

**Reply by tbnobody** [reply]
> OpenDTU initiates a connect request via an external MQTT client software (https://github.com/marvinroger/async-mqtt-client) to the MQTT server (e.g., ioBroker).

Thats not true. OpenDTU does not use this library.

It uses: https://github.com/bertmelis/espMqttClient

Please report your issue here to get it fixed in the upstream repository.

**Reply by yippieyaray** [reply]
Okay, I opened a case at https://github.com/bertmelis/espMqttClient

> "That's not true. OpenDTU does not use this library. It uses: https://github.com/bertmelis/espMqttClient"

Yes, but the library(https://github.com/marvinroger/async-mqtt-client) seems to be more or less the master of https://github.com/bertmelis/espMqttClient. In my opinion, reporting bugs in masters first is always the best strategy to propagate fixes to more than one fork/copy".
Hopefully, OpenDTU will not switch over to another mqtt client fork, for what reason ever. If it does switch, the same error might pop up again. But anyway, this master seems to be dead.
